File: src/helius_agent/agents/hitl.py
import logging
import os
from typing import Any, Dict, Optional

from langgraph.types import Command

logger = logging.getLogger(__name__)

# Secure CI/CD bypass token
HITL_BYPASS_ENV = os.getenv("HITL_BYPASS_TOKEN")

def handle_hitl_interrupt(agent: Any, config: Dict[str, Any], interactive: bool = True) -> Optional[Any]:
    """
    Processes pending Human-In-The-Loop (HITL) interrupts in the LangGraph state.
    
    Args:
        agent: The compiled LangGraph / DeepAgent.
        config: The LangGraph config dict (must contain thread_id).
        interactive: Whether to prompt via CLI if no bypass token is provided.
        
    Returns:
        The new graph state after resumption, or None if no interrupts were pending.
    """
    state = agent.get_state(config)
    
    # 1. Check if the graph is actually paused on an interrupt
    if not state.next or not state.tasks:
        return None
        
    task = state.tasks[0]
    if not task.interrupts:
        return None

    # 2. Extract the payload yielded by the tool's `interrupt({...})` call
    interrupt_payload = task.interrupts[0].value
    
    # Supports both our custom tools and generic action requests
    cmd_name = interrupt_payload.get("command") or interrupt_payload.get("name") or "Unknown Action"
    reason = interrupt_payload.get("reason") or "No justification provided by agent."

    logger.info("HITL Interrupt paused execution for tool: %s", cmd_name)

    # 3. Check for CI/CD Bypass Token via LangGraph's standard 'configurable' dict
    configurable = config.get("configurable", {})
    provided_token = configurable.get("hitl_bypass_token")

    if provided_token:
        if not HITL_BYPASS_ENV:
            logger.warning("Bypass requested but HITL_BYPASS_TOKEN is not set in environment.")
        elif provided_token == HITL_BYPASS_ENV:
            logger.info("✅ Bypass token validated. Auto-approving dangerous action.")
            return agent.invoke(Command(resume={"approved": True}), config)
        else:
            logger.warning("❌ Bypass token mismatch! Falling back to interactive mode.")

    # 4. Interactive CLI Approval
    if interactive:
        print("\n" + "="*60)
        print("🛑 HUMAN-IN-THE-LOOP AUTHORIZATION REQUIRED 🛑")
        print(f"Requested Action : {cmd_name}")
        print(f"Agent's Reason   : {reason}")
        print("="*60)
        
        user_in = input("\nApprove execution? (y/N): ").strip().lower()
        is_approved = user_in in ('y', 'yes')
        
        if is_approved:
            logger.info("Human approved the action.")
        else:
            logger.info("Human rejected the action.")
            
        # Resume the graph, passing the boolean decision back to the tool
        return agent.invoke(Command(resume={"approved": is_approved}), config)

    # 5. Non-interactive & No bypass -> Force Reject
    # We MUST resume the graph with a rejection. If we just return an error string, 
    # the graph remains indefinitely suspended and the agent loop breaks.
    logger.error("HITL required but interactive mode is off and no valid bypass token provided. Auto-rejecting.")
    return agent.invoke(Command(resume={"approved": False, "error": "interactive_mode_disabled"}), config)

__all__ =["handle_hitl_interrupt", "HITL_BYPASS_ENV"]

File: source/agents/hitl.ts
import readline from 'node:readline/promises';
import {stdin as input, stdout as output} from 'node:process';

import {Command} from '@langchain/langgraph';

import {getSettings} from '../utils/config.js';
import {getLogger} from '../utils/logger.js';
import {safeTokenCompare} from '../utils/security.js';

const logger = getLogger('agents.hitl');

export enum HITLAction {
	APPROVE = 'approve',
	EDIT = 'edit',
	REJECT = 'reject',
}

export class HITLDecision {
	action: HITLAction;
	editedCommand?: string;
	rejectionReason?: string;

	constructor(
		action: HITLAction,
		opts: {editedCommand?: string; rejectionReason?: string} = {},
	) {
		this.action = action;
		this.editedCommand = opts.editedCommand;
		this.rejectionReason = opts.rejectionReason;
	}

	get approved(): boolean {
		return this.action === HITLAction.APPROVE || this.action === HITLAction.EDIT;
	}

	toResumePayload(actionName: string, originalArgs: Record<string, unknown>): Record<string, unknown> {
		if (this.action === HITLAction.APPROVE) {
			return {decisions: [{type: 'approve'}]};
		}
		if (this.action === HITLAction.REJECT) {
			return {decisions: [{type: 'reject', feedback: this.rejectionReason ?? 'Rejected by human.'}]};
		}
		if (this.action === HITLAction.EDIT) {
			return {
				decisions: [{
					type: 'edit',
					editedAction: {
						name: actionName,
						args: {
							...originalArgs,
							command: this.editedCommand,
						},
					},
				}],
			};
		}
		return {decisions: [{type: 'reject'}]};
	}
}

export type HITLPresenter = {
	present(payload: Record<string, unknown>): Promise<HITLDecision> | HITLDecision;
};

export class CLIPresenter implements HITLPresenter {
	async present(payload: Record<string, unknown>): Promise<HITLDecision> {
		const actionName = String(payload.name ?? 'Unknown Action');
		const args = (payload.args as Record<string, unknown>) ?? payload;
		const cmd = String(args.command ?? actionName);
		const reason = String(
			payload.reason ?? `Action '${actionName}' requires approval.`,
		);

		const line = '='.repeat(62);
		const divider = '-'.repeat(62);
		console.log('');
		console.log(line);
		console.log('HUMAN AUTHORIZATION REQUIRED');
		console.log(line);
		console.log(`Action : ${cmd}`);
		console.log(`Reason : ${reason}`);
		console.log(divider);
		console.log('[y] Approve   [e] Edit command   [n] Reject  (default: n)');
		console.log(line);

		const rl = readline.createInterface({input, output});
		try {
			const choice = (await rl.question('  Your choice: ')).trim().toLowerCase();
			if (choice === 'y' || choice === 'yes') {
				logger.info(`HITL approve cmd=${cmd}`);
				return new HITLDecision(HITLAction.APPROVE);
			}
			if (choice === 'e' || choice === 'edit') {
				console.log(`\n  Current: ${cmd}`);
				const edited = (await rl.question('  Modified command (blank = approve as-is): '))
					.trim();
				if (edited) {
					logger.info(`HITL edit cmd=${cmd}`);
					return new HITLDecision(HITLAction.EDIT, {editedCommand: edited});
				}
				logger.info(`HITL approve cmd=${cmd}`);
				return new HITLDecision(HITLAction.APPROVE);
			}
			const rej = (await rl.question('  Rejection reason (optional): ')).trim();
			logger.info(`HITL reject cmd=${cmd}`);
			return new HITLDecision(HITLAction.REJECT, {
				rejectionReason: rej || 'Rejected by human.',
			});
		} catch {
			logger.warn(`HITL prompt interrupted cmd=${cmd}`);
			return new HITLDecision(HITLAction.REJECT, {
				rejectionReason: 'Prompt interrupted.',
			});
		} finally {
			rl.close();
		}
	}
}

export async function handleHitlInterrupt(
	agent: any,
	config: Record<string, unknown>,
	presenter?: HITLPresenter,
): Promise<any | null> {
	const settings = getSettings();

	const state = await agent.getState(config);
	if (!state?.next || state.next.length === 0) {
		return null;
	}
	if (!state.tasks || state.tasks.length === 0) {
		return null;
	}

	const task = state.tasks[0];
	if (!task.interrupts || task.interrupts.length === 0) {
		return null;
	}

	const payload = (task.interrupts[0]?.value ?? {}) as Record<string, unknown>;
	const args = (payload.args as Record<string, unknown>) ?? payload;
	const actionName = String(payload.name ?? 'Unknown');
	logger.info(`HITL requested action=${actionName}`);
	const configurable = (config as {configurable?: Record<string, unknown>})
		.configurable;
	const provided = configurable?.hitl_bypass_token as string | undefined;
	const expected = settings.hitl_bypass_token ?? undefined;

	if (provided && expected && safeTokenCompare(provided, expected)) {
		logger.info('HITL bypass token accepted');
		const decision = new HITLDecision(HITLAction.APPROVE);
		return new Command({resume: decision.toResumePayload(actionName, args)});
	}

	if (settings.hitl.interactive) {
		const activePresenter = presenter ?? new CLIPresenter();
		const decision = await activePresenter.present(payload);
		logger.info(`HITL decision=${decision.action}`);
		return new Command({resume: decision.toResumePayload(actionName, args)});
	}

	const decision
    = settings.hitl.default_action === 'approve'
    	? new HITLDecision(HITLAction.APPROVE)
    	: new HITLDecision(HITLAction.REJECT, {
    		rejectionReason: 'Non-interactive mode; no bypass token provided.',
    	});

	logger.info(`HITL default decision=${decision.action}`);
	return new Command({resume: decision.toResumePayload(actionName, args)});
}
